sdk/lib/rtl/priv.c

```diff
@@ -486,3 +486,97 @@ RtlAdjustPrivilege(IN ULONG Privilege,
 
     return STATUS_SUCCESS;
 }
+
+#if (NTDDI_VERSION >= NTDDI_VISTA)
+/**
+ * @brief
+ * Removes all privileges in the specified access token.
+ *
+ * @param[in] TokenHandle
+ * A handle to the access token that contains the privileges to be removed.
+ *
+ * @param[in] PrivilegesToKeep
+ * A pointer to an array of privilege values (defined as SE_XXX_PRIVILEGE) that specify
+ * the privileges to keep in the token.
+ *
+ * @param[in] PrivilegeCount
+ * Specifies the number of entries in the PrivilegesToKeep array.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if privileges removed successfully.
+ * STATUS_INVALID_PARAMETER is returned if input privilege value greater than
+ * SE_MAX_WELL_KNOWN_PRIVILEGE. STATUS_NOT_ALL_ASSIGNED is returned if The token does
+ * not have one or more of the privileges specified in the PrivilegesToKeep parameter,
+ * and no privileges were removed. A failure NTSTATUS code is returned otherwise.
+ */
+NTSTATUS
+NTAPI
+RtlRemovePrivileges(_In_ HANDLE TokenHandle,
+                    _In_reads_opt_(PrivilegeCount) _When_(PrivilegeCount != 0, _Notnull_) PULONG PrivilegesToKeep,
+                    _In_ ULONG PrivilegeCount)
+{
+    UINT64 PrivilegesToKeepBitmap;
+    ULONG i, ReturnLength;
+    UCHAR Buffer[sizeof(TOKEN_PRIVILEGES) +
+                 sizeof(LUID_AND_ATTRIBUTES) * (SE_MAX_WELL_KNOWN_PRIVILEGE - SE_MIN_WELL_KNOWN_PRIVILEGE)];
+    PTOKEN_PRIVILEGES Privileges;
+
+    DPRINT("RtlRemovePrivileges(%p, %p, %u)\n", TokenHandle, PrivilegesToKeep, PrivilegeCount);
+
+    /* Save privileges that should be keep */
+    PrivilegesToKeepBitmap = 0;
+    if (PrivilegeCount)
+    {
+        for (i = 0; i < PrivilegeCount; i++)
+        {
+            PrivilegesToKeepBitmap |= (1ULL << PrivilegesToKeep[i]);
+            if (PrivilegesToKeep[i] > SE_MAX_WELL_KNOWN_PRIVILEGE)
+            {
+                return STATUS_INVALID_PARAMETER;
+            }
+        }
+    }
+
+    /* Get token privileges information */
+    NTSTATUS Status = ZwQueryInformationToken(TokenHandle,
+                                              TokenPrivileges,
+                                              Buffer,
+                                              sizeof(Buffer),
+                                              &ReturnLength);
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    /* Remove all privileges that we don't need to keep */
+    Privileges = (PTOKEN_PRIVILEGES)Buffer;
+    for (i = 0; i < Privileges->PrivilegeCount; i++)
+    {
+        LARGE_INTEGER Privilege = *(LARGE_INTEGER*)&Privileges->Privileges[i].Luid;
+        if (PrivilegesToKeepBitmap & (1ULL << Privilege.QuadPart))
```

Comment:
Before these shifts: a `if (Privilege.QuadPart > SE_MAX_WELL_KNOWN_PRIVILEGE)` validation check similar to the one above seems to be missing.

Reply:
The first validation checks input parameter, and checks `SE_MAX_WELL_KNOWN_PRIVILEGE` only, doesn't check `SE_MIN_WELL_KNOWN_PRIVILEGE`.
Here the value returned by `ZwQueryInformationToken`, so the check in here is not necessary if `ZwQueryInformationToken` returns correct value which lower than `SE_MAX_WELL_KNOWN_PRIVILEGE`.
The logic above is the same with Windows.

Reply:
Hm OK, so maybe just an `ASSERT(Privilege.QuadPart <= SE_MAX_WELL_KNOWN_PRIVILEGE);` would be enough.
Also,
>  The logic above is the same with Windows.

Perhaps but you know, if said logic was "buggy" and then it turned out that sometime from now they issue a fix for (e.g.) this function, while we could have found said fix by ourselves much before.... It would have then been better to "improve" the logic from windows instead of making it "the same" (but buggy).